ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/NettyPerfClient.java

```diff
@@ -0,0 +1,501 @@
+      }
+    });
+    logger.info("Connecting to {}:{}", host, port);
+    b.remoteAddress(host, port);
+    perfClientStartTime = System.currentTimeMillis();
+    for (int i = 0; i < concurrency; i++) {
+      b.connect().addListener(channelConnectListener);
+    }
+    logger.info("Created {} channel(s)", concurrency);
+    logger.info("NettyPerfClient started");
+  }
+
+  /**
+   * Shuts down the NettyPerfClient.
+   */
+  protected void shutdown() {
+    logger.info("Shutting down NettyPerfClient");
+    isRunning = false;
+    group.shutdownGracefully();
+    long totalRunTimeInMs = System.currentTimeMillis() - perfClientStartTime;
+    try {
+      if (!group.awaitTermination(5, TimeUnit.SECONDS)) {
+        logger.error("Netty worker did not shutdown within timeout");
+      } else {
+        logger.info("NettyPerfClient shutdown complete");
+      }
+    } catch (InterruptedException e) {
+      logger.error("NettyPerfClient shutdown interrupted", e);
+    } finally {
+      logger.info("Executed for approximately {} s and sent {} requests ({} requests/sec)",
+          (float) totalRunTimeInMs / (float) Time.MsPerSec, totalRequestCount.get(),
+          (float) totalRequestCount.get() * (float) Time.MsPerSec / (float) totalRunTimeInMs);
+      Snapshot rttStatsSnapshot = perfClientMetrics.requestRoundTripTimeInMs.getSnapshot();
+      logger.info("RTT stats: Min - {} ms, Mean - {} ms, Max - {} ms", rttStatsSnapshot.getMin(),
+          rttStatsSnapshot.getMean(), rttStatsSnapshot.getMax());
+      logger.info("RTT stats: 95th percentile - {} ms, 99th percentile - {} ms, 999th percentile - {} ms",
+          rttStatsSnapshot.get95thPercentile(), rttStatsSnapshot.get99thPercentile(),
+          rttStatsSnapshot.get999thPercentile());
+      reporter.stop();
+      shutdownLatch.countDown();
+    }
+  }
+
+  /**
+   * Blocking function to wait on the NettyPerfClient shutting down.
+   * @throws InterruptedException
+   */
+  protected void awaitShutdown()
+      throws InterruptedException {
+    shutdownLatch.await();
+  }
+
+  /**
+   * Custom handler that sends out the request and receives responses.
+   */
+  private class RequestSender extends SimpleChannelInboundHandler<Object> {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private HttpRequest request;
+    private HttpResponse response;
+    private ChunkedInput<HttpContent> chunkedInput;
+
+    private int chunksReceived;
+    private long sizeReceived;
+    private long lastChunkReceiveTime;
+    private long requestStartTime;
+    private long requestId = 0;
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) {
+      perfClientMetrics.channelCreationRate.mark();
+      logger.trace("Channel {} active", ctx.channel());
+      sendRequest(ctx);
+    }
+
+    @Override
+    public void channelRead0(ChannelHandlerContext ctx, Object in) {
```

Comment:
`Object` is the signature in the parent, it can be refined to any child which is done in the other places. It doesn't really matter - but I will change this to `HttpObject` too.